mixer/tools/adapterlinter/main.go

```diff
@@ -0,0 +1,163 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"os"
+	"sort"
+	"strings"
+)
+
+var exitCode int
+
+var invalidImportPaths = map[string]string{
+	"log": "\"log\" import is not recommended; Adapters must instead use env.Logger for logging during execution. " +
+		"This logger understands about which adapter is running and routes the data to the place where the operator " +
+		"wants to see it.",
+	"github.com/golang/glog": "\"github.com/golang/glog\" import is not recommended; Adapters must instead use env.Logger for logging during execution. " +
+		"This logger understands about which adapter is running and routes the data to the place where the operator " +
+		"wants to see it.",
+}
+
+func main() {
+	flag.Parse()
+	var reports []string
+	if flag.NArg() == 0 {
+		reports = doAllDirs([]string{"."})
+	} else {
+		reports = doAllDirs(flag.Args())
+	}
+
+	for _, r := range reports {
+		reportErr(r)
+	}
+	os.Exit(exitCode)
+}
+
+func doAllDirs(args []string) []string {
+	reports := make([]string, 0)
+	for _, name := range args {
+		// Is it a directory?
+		if fi, err := os.Stat(name); err == nil && fi.IsDir() {
+			for _, r := range doDir(name) {
+				reports = append(reports, r.msg)
+			}
+		} else {
+			reportErr(fmt.Sprintf("not a directory: %s", name))
+		}
+	}
+	return reports
+}
+
+func doDir(name string) reports {
+	notests := func(info os.FileInfo) bool {
+		if !info.IsDir() && strings.HasSuffix(info.Name(), ".go") &&
+			!strings.HasSuffix(info.Name(), "_test.go") {
+			return true
+		}
+		return false
+	}
+	fs := token.NewFileSet()
+	pkgs, err := parser.ParseDir(fs, name, notests, parser.Mode(0))
+	if err != nil {
+		reportErr(fmt.Sprintf("%v", err))
+		return nil
+	}
+	rpts := make(reports, 0)
+	for _, pkg := range pkgs {
+		for _, r := range doPackage(fs, pkg) {
+			rpts = append(rpts, r)
+		}
+	}
+	sort.Sort(rpts)
+	return rpts
+}
+
+func doPackage(fs *token.FileSet, pkg *ast.Package) reports {
+	v := newVisitor(fs)
+	for _, file := range pkg.Files {
+		ast.Walk(&v, file)
+	}
+	return v.reports
+}
+
+func newVisitor(fs *token.FileSet) visitor {
+	return visitor{
+		fs: fs,
+	}
+}
+
+type visitor struct {
+	reports reports
+	fs      *token.FileSet
+}
+
+func (v *visitor) Visit(node ast.Node) ast.Visitor {
+	if node == nil {
+		return nil
+	}
+
+	switch d := node.(type) {
+	case *ast.GoStmt:
+		v.reports = append(v.reports,
+			report{
+				d.Pos(),
+				fmt.Sprintf("%v:%v:%v:Adapters must use env.ScheduleWork or env.ScheduleDaemon in order to "+
```

Comment:
nit: worth extracting into a separate method to avoid the massive indent? 

```
case *ast.GoStmt: v.reports = append (v.reports, goroutineReport(d.Pos()))
case *ast.ImportSpect:
         if d.Path == nil {
             continue
         }
         p := strings.Trim(...)
         for badImp, reportMsg := range invalidImportPaths {
                 if p != badImp {
                    continue
                 }
                 v.reports = append(v.reports, invalidImportReport(d.Path.Pos())
         }
}
```